### PR TITLE
Add Cleaner task to clean completed invocations

### LINKED ISSUE #1714: Action::ScheduleInvocationStatusCleanup can get lost on recoveries
I think that the system is susceptible to losing `Action::ScheduleInvocationStatusCleanup` messages that are used to communicate to replicas that a completed invocation result should be cleaned up at some point in the future. We do generate the `Action::ScheduleInvocationStatusCleanup` when we complete an invocation. After writing the completed invocation status to RocksDB, we try to self-propose this message. If the PP loses leadership just before it can self-propose, this message will never be seen by the replicas, because the new leader does not check whether all completed invocations have a cleanup timer registered.

A possible solution is to let the new leader check whether there are completed invocations that don't have a cleanup timer registered (e.g. by storing a flag in the completed invocation that signals this). Alternatively, we don't rely on individual timers for cleaning up completed invocations.

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Fix #1714, based on https://github.com/restatedev/restate/pull/1843

### CODE REVIEW

File: crates/worker/src/partition/state_machine/tests/idempotency.rs [hunk 1]
```diff
@@ -142,6 +142,126 @@ async fn start_idempotent_invocation() {
...
+        .get_invocation_status(&invocation_id)
+        .await
+        .unwrap();
+    let_assert!(InvocationStatus::Completed(completed_invocation) = invocation_status);
+    assert_eq!(
+        completed_invocation.response_result,
+        ResponseResult::Success(response_bytes)
+    );
+    assert!(unsafe { completed_invocation.timestamps.completed_transition_time() }.is_some());
+    assert_eq!(completed_invocation.completion_retention, retention);
```
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/tests/idempotency.rs:262]: Should we also test that the clean up happens? If you start the test with `start_paused = true`, then tokio will advance the time to the next event if nothing else can run. That way we can simulate the hourly cleanup in basically no time. Here are a few more details https://docs.rs/tokio/latest/tokio/time/fn.pause.html.

File: crates/worker/src/partition/cleaner.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+        Ok(())
+    }
+
+    pub(super) async fn do_cleanup(
+        storage: &Storage,
+        bifrost: &Bifrost,
+        partition_key_range: RangeInclusive<PartitionKey>,
+        bifrost_envelope_source: &Source,
+    ) -> anyhow::Result<()> {
```
Comment (tillrohrmann) [inline crates/worker/src/partition/cleaner.rs:104]: Maybe add debug statement that the cleanup is happening. This could help figuring out if the system becomes slow that this might be caused by the extra RocksDB load.

File: crates/worker/src/partition/cleaner.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        debug!(restate.node = %node_id, restate.partition.id = %partition_id, "Running cleaner");
+
+        let bifrost_envelope_source = Source::Processor {
+            partition_id,
+            partition_key: None,
+            leader_epoch,
+            node_id,
+        };
+
+        let mut interval = tokio::time::interval(cleanup_interval);
```
Comment (tillrohrmann) [inline crates/worker/src/partition/cleaner.rs:82]: I would set `MissedTickBehavior::Delay` so that slow clean up operations won't result into burst calls and therefore even more load.

**Review by tillrohrmann — APPROVED** [review]
Thanks a lot for fixing the problem with cleaning up completed invocations @slinkydeveloper. The current solution halves the number of clean up messages needed for invocations that we retain. In the future we might be able to further reduce it by propagating only the time through Bifrost and then let every replica do the scanning themselves.

The changes look good to me. I had a few minor comments. +1 for merging after resolving them.

File: crates/storage-api/proto/dev/restate/storage/v1/domain.proto [hunk 4]
```diff
@@ -112,15 +112,21 @@ message NeoInvocationStatus {
     InvocationTarget invocation_target = 2;
     Source source = 3;
     SpanContext span_context = 4;
+    repeated ServiceInvocationResponseSink response_sinks = 7;
+    Duration completion_retention_time = 11;
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/storage-api/proto/dev/restate/storage/v1/domain.proto:116]: it is a duration indeed, let me fix this name.

File: crates/storage-api/src/invocation_status_table/mod.rs [hunk 5]
```diff
@@ -471,7 +605,7 @@ mod test_util {
                 journal_metadata: JournalMetadata::initialize(ServiceInvocationSpanContext::empty()),
                 pinned_deployment: None,
                 response_sinks: HashSet::new(),
-                timestamps: StatusTimestamps::now(),
+                timestamps: StatusTimestamps::init(MillisSinceEpoch::now()),
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/storage-api/src/invocation_status_table/mod.rs:608]: i marked it unsafe for now, so in the tests i simply changed it with this to avoid the unsafe noise

File: crates/storage-api/src/invocation_status_table/mod.rs [hunk 6]
```diff
@@ -35,40 +35,154 @@ pub enum SourceTable {
...
+        }
     }
 
     /// Update the statistics with an updated [`Self::modification_time()`].
-    pub fn update(&mut self) {
+    ///
+    /// # Safety
+    /// The value of this time is not consistent across replicas of a partition, because it's not agreed.
+    /// You **MUST NOT** use it within the Partition processor business logic, but only for observability purposes.
+    pub unsafe fn update(&mut self) {
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/storage-api/src/invocation_status_table/mod.rs:80]: ok will remove unsafe for write methods and the init/new

File: crates/worker/src/partition/cleaner.rs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        loop {
+            tokio::select! {
+                _ = interval.tick() => {
+                    Self::do_cleanup(&storage, &bifrost, partition_key_range.clone(), &bifrost_envelope_source).await?
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:86]: > (e.g. failed to produce to Bifrost)

Aren't "failed to produce to Bifrost" or "cannot scan invocations" catastrophic problems anyway?

File: crates/worker/src/partition/state_machine/tests/idempotency.rs (hunk 1 above)
Comment (slinkydeveloper) [reply crates/worker/src/partition/state_machine/tests/idempotency.rs:262]: I think what I would rather do is have such test in the cleaner, to check if the PurgeInvocation commands are produced to bifrost.

File: crates/worker/src/partition/cleaner.rs (hunk 3 above)
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:82]: :+1:

File: crates/worker/src/partition/state_machine/tests/idempotency.rs [hunk 8]
```diff
@@ -142,6 +142,126 @@ async fn start_idempotent_invocation() {
     );
 }
 
+#[test(tokio::test(flavor = "multi_thread", worker_threads = 2))]
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/state_machine/tests/idempotency.rs:145]: Will tackle this in a followup for all the tests.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -99,6 +99,7 @@ jobs:
       restateCommit: ${{ github.event.pull_request.head.sha || github.sha }}
       envVars: |
         RESTATE_WORKER__EXPERIMENTAL_FEATURE_NEW_INVOCATION_STATUS_TABLE=true
+        RESTATE_WORKER__CLEANUP_INTERVAL=1s
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply .github/workflows/ci.yml:102]: Nice it works, gonna apply this to the test tool.

**Review by tillrohrmann — COMMENTED** [review]
Thanks for updating the PR @slinkydeveloper. The test looks good to me. I had a few minor questions. Please take a look.

File: crates/worker/src/partition/cleaner.rs [hunk 10]
```diff
@@ -161,3 +161,150 @@ where
...
+        ) -> impl Stream<Item = restate_storage_api::Result<(InvocationId, InvocationStatus)>> + Send
+        {
+            stream::iter(self.0.clone()).map(Ok)
+        }
+    }
+
+    // Start paused makes sure the timer is immediately fired
+    #[test(tokio::main(flavor = "current_thread", start_paused = true))]
+    pub async fn cleanup_works() {
+        // set numbers of partitions to 1 to easily find all sent messages by the shuffle
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:217]: lol you can see how badly i copy pasted this code :D

File: crates/worker/src/partition/cleaner.rs [hunk 11]
```diff
@@ -161,3 +161,150 @@ where
...
+                    ..CompletedInvocation::mock_neo()
+                }),
+            ),
+            (
+                not_expired_invocation_2,
+                InvocationStatus::Completed(CompletedInvocation {
+                    completion_retention_duration: Duration::ZERO,
+                    timestamps: StatusTimestamps::now(),
+                    ..CompletedInvocation::mock_neo()
+                }),
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:263]: `StatusTimestamps::now()` won't set the `completed_transition_time()`, let me add a comment.

File: crates/worker/src/partition/cleaner.rs [hunk 12]
```diff
@@ -161,3 +161,150 @@ where
...
+            &self,
+            _: RangeInclusive<PartitionKey>,
+        ) -> impl Stream<Item = restate_storage_api::Result<(InvocationId, InvocationStatus)>> + Send
+        {
+            stream::iter(self.0.clone()).map(Ok)
+        }
+    }
+
+    // Start paused makes sure the timer is immediately fired
+    #[test(tokio::main(flavor = "current_thread", start_paused = true))]
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:215]: true but the macro requires it when used in combination with start_paused:

```
error: The `start_paused` option requires the `current_thread` runtime flavor. Use `#[tokio::main(flavor = "current_thread")]`
   --> crates/worker/src/partition/cleaner.rs:215:39
    |
215 |     #[test(tokio::main(start_paused = true))]
    |                                       ^^^^
```

File: crates/worker/src/partition/cleaner.rs [hunk 13]
```diff
@@ -161,3 +161,150 @@ where
...
+        let not_expired_invocation_2 =
+            InvocationId::from_parts(PartitionKey::MIN, InvocationUuid::new());
+        let not_completed_invocation =
+            InvocationId::from_parts(PartitionKey::MIN, InvocationUuid::new());
+
+        let mock_storage = MockInvocationStatusReader(vec![
+            (
+                expired_invocation,
+                InvocationStatus::Completed(CompletedInvocation {
+                    completion_retention_duration: Duration::ZERO,
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:246]: Yes, but I set it to Zero to be sure that the test passes. If I set it to a higher value, the test might not pass b/c the now() + duration needs to be smaller or equal than when do_cleaner calls now()

File: crates/worker/src/partition/cleaner.rs [hunk 14]
```diff
@@ -161,3 +161,150 @@ where
...
+        let mut log_entries = bifrost.read_all(partition_id.into()).await.unwrap();
+        let bifrost_message = log_entries
+            .remove(0)
+            .try_decode::<Envelope>()
+            .unwrap()
+            .unwrap();
+
+        assert_that!(
+            bifrost_message.command,
+            pat!(Command::PurgeInvocation(pat!(PurgeInvocationRequest {
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:304]: I just figured that we don't. Let me add a test

File: crates/worker/src/partition/cleaner.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/worker/src/partition/cleaner.rs:215]: I think the problem is `tokio::main`. It should be `tokio::test` and then it should work. Ref: https://github.com/restatedev/restate/blob/b456f2c4d5879cb6aff48da1c8f827c43879309d/crates/core/src/network/connection_manager.rs#L906.

File: crates/worker/src/partition/cleaner.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/worker/src/partition/cleaner.rs:245]: With `start_paused` the cleaner should eventually (acutally quite quickly) arrive at the expiration time. Isn't this the purpose of the test that we test that something that is supposed to expire in the future will actually cleaned up by then?

File: crates/worker/src/partition/cleaner.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/worker/src/partition/cleaner.rs]: What does this mean for the invocation status on a higher/logical level?

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs]: it practically means the invocation was created with the old table.

File: crates/worker/src/partition/cleaner.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:245]: Yes but i still get the time with `SystemTime::now()` from std with the cleaner loop, and that is not affected by `start_paused = true`

File: crates/worker/src/partition/cleaner.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:215]: well changing to main has some other interesting effect...

```
thread 'partition::cleaner::tests::cleanup_works' panicked at crates/worker/src/partition/cleaner.rs:271:12:
spawn_child called outside of a task-center task
```

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:215]: same doesn't happen with `tokio::main`...

File: crates/worker/src/partition/cleaner.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs]: I simplified this code to make it clear this is for old invocations.

File: crates/worker/src/partition/cleaner.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:215]: and if i use `run_in_scope` it won't work because i need to poll the future somehow

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:215]: i also tried to wrap all the test in `run_in_scope` and it won't work either...

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/worker/src/partition/cleaner.rs:215]: I think you should not call `spawn_child` outside of a TaskCenter task. If you change it to `spawn`, then it works.

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:215]: That solves it

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/worker/src/partition/cleaner.rs:215]: And it "works" with `tokio::main` because with `#[test(tokio::main(flavor = "current_thread", start_paused = true))]` the test won't be executed.

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/worker/src/partition/cleaner.rs:215]: When using `spawn` and `tokio::test`, the test fails with

```
thread 'partition::cleaner::tests::cleanup_works' panicked at crates/worker/src/partition/cleaner.rs:293:14:
removal index (is 0) should be < len (is 0)
```

File: crates/worker/src/partition/cleaner.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/worker/src/partition/cleaner.rs:245]: I see. So what we would need is a clock that we can control in our test. Maybe something like https://docs.rs/quanta/0.12.3/quanta/struct.Clock.html helps here?

File: crates/worker/src/partition/cleaner.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs]: solved

File: crates/worker/src/partition/cleaner.rs [hunk 15]
```diff
@@ -281,6 +281,9 @@ mod tests {
         )
         .unwrap();
 
+        // By yielding once we let the cleaner task run, and perform the cleanup
+        tokio::task::yield_now().await;
```
Comment (tillrohrmann) [inline crates/worker/src/partition/cleaner.rs:285]: What I don't like about this test is that it depends on the fact that `Interval` immediately triggers and that the task has been registered when we reach this point. Maybe it is what it is but if we had the clock used by the cleaner under our control, then we could make the test more robust.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating the PR @slinkydeveloper. I think the test could be a bit more robust and testing a non-zero retention duration but maybe this is a follow up.

File: crates/worker/src/partition/cleaner.rs (hunk 15 above)
Comment (slinkydeveloper) [reply crates/worker/src/partition/cleaner.rs:285]: I think this is ok for the time being
